Ability to add custom dashboard actions

mweimerskirch opened the pull request:
The actions displayed on the dashboard for each admin type are hard-coded: only "create" and "list" are available. I wanted to customize them to add direct links to special views. Unfortunately, that was not possible yet, so I came up with a solution. The templates for the two existing actions are moved into individual templates: "dashboard__action_list.html.twig" and "dashboard__action_create.html.twig". What remains is a "for" loop that retrieves the necessary configuration from the getDashboardActions methon in the admin class. This method can then be overwritten for individual types.

This is not a finished patch yet, as I first wanted to gather feedback: Would this patch be accepted if I finish it? Is this something other people would find useful as well?

Koc commented:
fixes #2953

Koc commented:
@mweimerskirch looks nice. What about adding default template and options like `icon_class`. Sometimes no any logic needed in the template and we can reuse default template without creating new one

Koc commented on Admin/Admin.php:
```diff
@@ -2886,4 +2886,27 @@ public function checkAccess($action, $object = null)
         }
     }
+
+    public function getDashboardActions()
```
Maybe you need update `AdminInterface` also. //cc @rande, @Soullivaneuh

core23 replied:
This needs to be done in a seperate PR. Otherwise it would be a BC break.

core23 replied:
Please add PHPDoc

mweimerskirch commented:
I worked a bit more on this idea:
* Added some PHPDoc.
* Added a check to see if there are any dashboard actions (previously, there was only a check to see if a "list" and "create" action were available.
* I also added a default template as @Koc suggested. I'm not 100% sure about this yet though.

Koc commented on Resources/views/Block/block_admin_list.html.twig:
```diff
@@ -27,40 +27,16 @@ file that was distributed with this source code.
-                                                            </a>
-                                                        {% endif %}
+                                                        {% for action in admin.dashboardActions %}
+                                                            {% include action.template with {'action': action} %}
```
`action.template|default('SonataAdminBundle:CRUD:dashboard__action.html.twig')`

mweimerskirch replied:
@Koc good idea, thanks

Koc commented on Resources/views/Block/block_admin_list.html.twig:
```diff
@@ -27,40 +27,16 @@ file that was distributed with this source code.
-                                                            </a>
-                                                        {% endif %}
+                                                        {% for action in admin.dashboardActions %}
+                                                            {% include action.template|default('SonataAdminBundle:CRUD:dashboard__action.html.twig') with {'action': action} %}
```
don't know, but maybe better use `include` function https://github.com/twigphp/Twig/issues/1899

mweimerskirch replied:
The issue you are referring to is still open, so no decision has been made yet. {% include %} is used everywhere in the code so far, so I think it might be better keeping it for now. If and when it is deprecated in Twig, it could be replaced everywhere in a single commit.

Koc commented:
Can you please add docs which explain this feature?

mweimerskirch commented:
I made the "translation_domain" key optional as well, added some documentation and squashed the commits.

mweimerskirch commented:
Anything else I can do to help this get merged?

OskarStark commented:
@mweimerskirch :+1: can you please check the CS? travis is not happy :smile: 

is it possible to add some tests?

Koc commented:
@mweimerskirch you should align spaces inside array:

```php
'label'              => 'link_add',
'translation_domain' => 'SonataAdminBundle',
```

core23 commented:
Please rebase and add some tests :+1:

core23 commented:
ping @mweimerskirch

mweimerskirch commented:
@core23 I can rebase over the weekend, but I unfortunately don't have time to add (meaningful) tests :-S

greg0ire commented:
It's great to have meaningful tests, but we can't have them, having at least some code coverage is great (it helps us get deprecation notices while running the tests, for instance).

greg0ire commented:
@mweimerskirch : can you rebase ?

core23 commented:
Ping @mweimerskirch

mweimerskirch commented:
I still haven't had the time, sorry. Too much work and I'm just human.
I will go through my open pull requests at the Symfony Live Cologne Hack Day (April 30th).

mweimerskirch commented:
@core23 I rebased the patch and added a test case.

greg0ire commented on Resources/views/CRUD/dashboard__action.html.twig:
```diff
@@ -0,0 +1,4 @@
+<a class="btn btn-link btn-flat" href="{{ action.url }}">
+    <i class="fa fa-{{ action.icon }}"></i>
+    {{ action.label|trans({}, action.translation_domain|default('SonataAdminBundle')) }}
+</a>
```
Please configure your editor to add newline at end of file, it should make the red pictogram disappear.

mweimerskirch replied:
@greg0ire the patch has already been merge. I'll watch out for this next time though

greg0ire replied:
Woops, right, my bad.
